File: architecture-skills/design-patterns/examples/flyweight/flyweight.go
// Flyweight pattern in Go.
//
// The factory hands out *TreeType pointers from a map cache, so every
// oak/green tree shares one struct. Immutability of the shared state is
// by convention in Go: nothing mutates a TreeType after creation.

package main

import "fmt"

// TreeType is the flyweight: intrinsic state shared by many trees.
type TreeType struct {
	Name  string
	Color string
}

func (t *TreeType) Draw(x, y int) {
	fmt.Printf("%s %s at (%d, %d)\n", t.Color, t.Name, x, y)
}

type TreeTypeFactory struct {
	cache map[string]*TreeType
}

func NewTreeTypeFactory() *TreeTypeFactory {
	return &TreeTypeFactory{cache: map[string]*TreeType{}}
}

func (f *TreeTypeFactory) Get(name, color string) *TreeType {
	key := name + "/" + color
	if t, ok := f.cache[key]; ok {
		return t
	}
	t := &TreeType{Name: name, Color: color}
	f.cache[key] = t
	return t
}

func (f *TreeTypeFactory) Len() int { return len(f.cache) }

// Tree is the context: extrinsic state plus the shared flyweight.
type Tree struct {
	X, Y int
	Kind *TreeType
}

func main() {
	factory := NewTreeTypeFactory()
	forest := []Tree{
		{1, 1, factory.Get("oak", "green")},
		{2, 4, factory.Get("pine", "dark green")},
		{5, 2, factory.Get("oak", "green")},
		{6, 6, factory.Get("birch", "light green")},
		{8, 3, factory.Get("pine", "dark green")},
		{9, 9, factory.Get("oak", "green")},
	}

	for _, tree := range forest {
		tree.Kind.Draw(tree.X, tree.Y)
	}

	fmt.Printf("trees planted: %d, tree types in memory: %d\n", len(forest), factory.Len())
}

File: architecture-skills/design-patterns/examples/flyweight/flyweight.swift
// Flyweight pattern in Swift.
//
// The flyweight must be a class: reference semantics let many trees
// share one instance, where a struct would be copied on every
// assignment and defeat the pattern. The stored properties are `let`,
// so the shared state is immutable.

/// Flyweight: immutable intrinsic state shared by many trees.
final class TreeType {
    let name: String
    let color: String

    init(name: String, color: String) {
        self.name = name
        self.color = color
    }

    func draw(x: Int, y: Int) {
        print("\(color) \(name) at (\(x), \(y))")
    }
}

final class TreeTypeFactory {
    private var cache: [String: TreeType] = [:]

    func get(name: String, color: String) -> TreeType {
        let key = "\(name)/\(color)"
        if let existing = cache[key] {
            return existing
        }
        let type = TreeType(name: name, color: color)
        cache[key] = type
        return type
    }

    var count: Int { cache.count }
}

/// Context: extrinsic state plus a reference to the shared flyweight.
struct Tree {
    let x: Int
    let y: Int
    let kind: TreeType
}

let factory = TreeTypeFactory()
let forest: [Tree] = [
    Tree(x: 1, y: 1, kind: factory.get(name: "oak", color: "green")),
    Tree(x: 2, y: 4, kind: factory.get(name: "pine", color: "dark green")),
    Tree(x: 5, y: 2, kind: factory.get(name: "oak", color: "green")),
    Tree(x: 6, y: 6, kind: factory.get(name: "birch", color: "light green")),
    Tree(x: 8, y: 3, kind: factory.get(name: "pine", color: "dark green")),
    Tree(x: 9, y: 9, kind: factory.get(name: "oak", color: "green")),
]

for tree in forest {
    tree.kind.draw(x: tree.x, y: tree.y)
}

print("trees planted: \(forest.count), tree types in memory: \(factory.count)")
